astropy/utils/data.py

```diff
@@ -1375,15 +1714,103 @@ def get_cached_urls():
     -------
     cached_urls : list
         List of cached URLs.
-    """
-    # The code below is modified from astropy.utils.data.download_file()
-    try:
-        dldir, urlmapfn = _get_download_cache_locs()
-    except OSError as e:
-        msg = 'Remote data cache could not be accessed due to '
-        estr = '' if len(e.args) < 1 else (': ' + str(e))
-        warn(CacheMissingWarning(msg + e.__class__.__name__ + estr))
-        return False
 
-    with shelve.open(urlmapfn) as url2hash:
+    See Also
+    --------
+    cache_contents : obtain a dictionary listing everything in the cache
+    """
+    with _cache() as (dldir, url2hash):
         return list(url2hash.keys())
+
+
+def cache_contents():
+    """Obtain a dict mapping cached URLs to filenames."""
+    with _cache() as (dldir, url2hash):
+        return ReadOnlyDict((k, os.path.join(dldir, v))
+                            for (k, v) in url2hash.items())
+
+
+_cache_zip_index_name = "index.json"
```

Comment:
Why can't this be a function input parameter?

From the code, it looks like the functions would create this file in the working directory and there is no way the user can change this behavior. Is that desirable? Shouldn't it live where the ZIP file is/will be?

Reply:
Actually it never exists in the filesystem at all - the contents are created as a string in memory and then written into the ZIP file under this name. The reason it isn't a parameter is that I want all ZIP files ever created to use the same name for this - you need to know the name the creator used for this when you are trying to read a ZIP file.

The key problem is that every file needs to be associated with a URL, and the ZIP file needs to record this association somehow. Upon reflection it's cleaner to just store each file under an appropriately quoted version of its URL; the ZIP file limit on length for archive member names is somewhere over 60_000 bytes. I think this is fine. See new implementation that doesn't need an index.